mesonbuild/mformat.py

```diff
@@ -975,9 +975,12 @@ def run(options: argparse.Namespace) -> int:
 
     sources: T.List[Path] = options.sources.copy() or [Path(build_filename)]
     if not options.configuration:
-        default_config_path = sources[0].parent / 'meson.format'
-        if default_config_path.exists():
-            options.configuration = default_config_path
+        current_path = Path.cwd()
```

Comment:
Searching from cwd is wrong, as you could invoke the command from an unrelated directory.

An option could be to search in parents of the given source file. My concern however is that you have no way to know when you reached the sources root.

Reply:
First, when searching upwards and encountering `meson.format`, the search should stop. If it is not found, I can think of two methods to determine the source root.

1. Use some heuristic methods, such as stopping the search when encountering files like `.git`. This method is the simplest and less accurate, but it is at least better than the current approach of only looking for `meson.format` in `sources[0].parent`.
2. First, search for `meson.build` files in each parent directory. If found, parse the file to check if it is the top-level `meson.build` (determined by whether it starts with `project()`). In my understanding, this method is accurate but less efficient.

Reply:
Iterating the parents of the current source is probably the way to go, even if we could potentially go outside the source root.

Reply:
This could lead to a problem: consider a repository A that does not contain a meson.format file. If it exists as a submodule within a repository B that does contain a meson.format file, running meson format on A would use the meson.format from repository B. This might not align with the user's intentions.

But at least it's better than only use `meson.format` in `sources[0].parent`. I'll update this PR.

Reply:
Done.

Reply:
> First, search for `meson.build` files in each parent directory. If found, parse the file to check if it is the top-level `meson.build` (determined by whether it starts with `project()`). In my understanding, this method is accurate but less efficient.

That is what we do in mesonbuild/interpreterbase/interpreterbase.py, see sanity_check_ast. But it does require instantiating `mparser.Parser()` once for each file...